x/wasm/internal/keeper/keeper.go

```diff
@@ -443,6 +447,7 @@ func (k Keeper) migrate(ctx sdk.Context, contractAddress sdk.AccAddress, caller
 // another native Go module directly. Thus, the keeper doesn't place any access controls on it, that is the
 // responsibility or the app developer (who passes the wasm.Keeper in app.go)
 func (k Keeper) Sudo(ctx sdk.Context, contractAddress sdk.AccAddress, msg []byte) (*sdk.Result, error) {
+	defer telemetry.MeasureSince(time.Now(), "wasm", "contract", "sudo")
```

Comment:
Nice, so we get a count and time used for all the keeper calls.

If we have recursive contract calls (A calls B calls C), we will get 3 prometheus events, right? (A + B + C, B + C, and C). Is it possible to annotate that these are nested somehow?

Reply:
These are very broad metrics that measure an operation but not link them with the contract itself. With recursive calls we would get 3 events. I had chosen to not add contract labels as this would be a growing dataset. But I agree that some sort of tracing makes a lot of sense. We should look into https://opentelemetry.io/ and friends instead.

Reply:
`pinned` could be a good category label in a new iteration